### DISCUSSION TITLE
Inconsistent line formatting when using ruff

### DISCUSSION THREAD

**Discussion by bavo96** [astral-sh/ruff#13264] [Q&A] [answered]
When using ruff formatter, I get an annoying behaviour like this:

First time format the code:
```python
-        support_care_config[profile]["AIConfig"][
-            "additional_fallback"
-        ] = additional_fallback
+        support_care_config[profile]["AIConfig"]["additional_fallback"] = (
+            additional_fallback
+        )
```

Change some code below this block and code, and then format the code again:
```python
-        support_care_config[profile]["AIConfig"]["additional_fallback"] = (
-            additional_fallback
-        )
+        support_care_config[profile]["AIConfig"][
+            "additional_fallback"
+        ] = additional_fallback
```

What might be the issue of this behavior? Really need some help here :(

My ruff is at version `0.5.3` and I don't have any `ruff.toml` file, using the default settings of ruff.

**Comment by MichaReiser** [comment]
Hmm, I'm unable to reproduce this [in the playground](https://play.ruff.rs/bde2fa87-1b04-4bb4-aedf-08676c0c439d?secondary=Format)

Can you tell me more about how you run ruff? Do you use Ruff in the editor or in the CLI? Do you have a more complete example that you could share?

**Reply by bavo96** [reply]
I'm currently using ruff via [neovim](https://github.com/neovim/neovim), calling ruff using [nvim-lspconfig](https://github.com/neovim/nvim-lspconfig).

This is my config for ruff (return an empty table):
```lua
elseif server == 'ruff' then -- ===> Linting, Formatting, Organization Imports in python
        -- https://docs.astral.sh/ruff/editors/#language-server-protocol
        return {
        }
```

This is where I load ruff:
```lua
        local servers = { 'pylsp', 'ruff', 'lua_ls', 'bashls', 'html', 'cssls' };

        for _, server in ipairs(servers) do
            require("lspconfig")[server].setup(
                {
                    settings = lsputils.settings(server),
                    on_attach = lsputils.on_attach,
                    capabilities = lsputils.capabilities(),
                    init_options = lsputils.init_options(server),
                }

            )
        end;
```
And I install ruff using `pip install ruff==0.5.3`.

I have a config to auto format on save like this:

```lua
    -- Auto formatting when saving file
    vim.api.nvim_create_autocmd("BufWritePre", {
        buffer = buff,
        callback = function()
            vim.lsp.buf.format { async = false }
        end
    })
```

**Reply by MichaReiser** [reply]
Hmm, I'm not very familiar with neovim. Maybe @dhruvmanila spots something? I suspect that you might be using two different formatters.

**Reply by bavo96** [reply]
Currently I have these lsp servers running:
<img width="821" alt="image" src="https://github.com/user-attachments/assets/24621ee0-5ff7-4cd5-b5fe-13df82c38bf3">
Do you think maybe the pylsp is overlapping with ruff?

Currently I only have this config to handle between ruff and pylsp:
```lua
function M.on_attach(server, buff)
    -- Disable hover in favor of python lsp server
    if server.name == 'ruff' then
        server.server_capabilities.hoverProvider = false
    end
```

**Reply by MichaReiser** [reply]
Did you install any of the plugins listed [here](https://github.com/python-lsp/python-lsp-server?tab=readme-ov-file#3rd-party-plugins)?

**Reply by dhruvmanila** [reply, marked as answer]
> Do you think maybe the pylsp is overlapping with ruff?

This could happen. The snippet that you've provided would send the formatting request to all the server that supports formatting. What you want instead is to only send it to a specific server which can be done by:

```lua
-- Auto formatting when saving file
vim.api.nvim_create_autocmd("BufWritePre", {
    buffer = buff,
    callback = function()
        vim.lsp.buf.format { 
			async = false,
			filter = function(client)
				return client.name == "ruff"
			end
		}
    end
})
```

**Reply by bavo96** [reply]
This is what I'm currently installing using `pip install -r requirements.txt`: 
```python
ruff==0.5.3
python-lsp-server==1.11.0
isort==5.13.2
```

**Reply by bavo96** [reply]
> > Do you think maybe the pylsp is overlapping with ruff?
> 
> This could happen. The snippet that you've provided would send the formatting request to all the server that supports formatting. What you want instead is to only send it to a specific server which can be done by:
> 
> ```lua
> -- Auto formatting when saving file
> vim.api.nvim_create_autocmd("BufWritePre", {
>     buffer = buff,
>     callback = function()
>         vim.lsp.buf.format { 
>       async = false,
>       filter = function(client)
>         return client.name == "ruff"
>       end
>     }
>     end
> })
> ```

I have turned off all linters and formatters from pylsp like this, but somehow the pylsp formatters and linters are still running:
```lua
    elseif server == 'pylsp' then -- ===> Completions, Definitions, Hover, References, Signature Help, and Symbols in Python
        return {
            pylsp = {
                plugins = {
                    -- Turn these off to use flake8, see: https://github.com/python-lsp/python-lsp-server
                    pycodestyle = {
                        enabled = false,
                    },
                    mccabe = { enabled = false },
                    pyflakes = { enabled = false },

                    -- Turn these off to use black, see: https://github.com/python-lsp/python-lsp-black
                    autopep8 = { enabled = false },
                    yapf = { enabled = false },

                    -- Currenlty turn off to use ruff
                    -- Linter
                    flake8 = { enabled = false },
                },
                configurationSources = {},
            }
        }
```
But after following your code, the formatting process is super fast now (maybe only using ruff), and the code doesn't change anymore :D I'll test more to see whether this issue has gone or not, and get back to confirm later :D. Thank you guys for your super fast library and super fast replies also ❤️

**Reply by bavo96** [reply]
After testing for a few days, the issue has gone <3 Many thanks for your fantastic solution @MichaReiser @dhruvmanila
